File: algorithm/Algorithm/basic/Code_day08_PrintAllPermutations.js

/**
 * 打印字符串的全排列
 * 允许重复
 */
function printAllPermutations1(str) {
    function process1(i) {
        if (i == str.length) {
            console.log(arr.join());
            return;
        }
        for (let j = i; j < arr.length; j++) {
            swap(arr, i, j);
            process1(i + 1);
            swap(arr, i, j);
        }
    }
    let arr = str.split('');
    process1(0);
}

/**
 * 打印全排列，不允许重复
 * 这里可以用hashset去重
 */
function printAllPermutations2(str) {
    function process2(i) {
        if (i == arr.length) {
            console.log(arr.join());
            return;
        }
        let set = new Set();
        for (let j = i; j < arr.length; j++) {
            if (!set.has(arr[j])) {
                set.add(arr[j]);
                swap(arr, i, j);
                process2(i + 1);
                swap(arr, i, j);
            }
        }
    }
    let arr = str.split('');
    process2(0);
}

function swap(arr, i, j) {
    let temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}


// for test
function main() {
    let test1 = "abc";
    printAllPermutations1(test1);
    console.log("======");
    printAllPermutations2(test1);
    console.log("======");

    let test2 = "acc";
    printAllPermutations1(test2);
    console.log("======");
    printAllPermutations2(test2);
    console.log("======");
}

main();

File: algorithm/Algorithm/basic/Code_day08_PrintAllPermutations.java
package class_08;

import java.util.HashSet;

public class Code_04_Print_All_Permutations {

	public static void printAllPermutations1(String str) {
		char[] chs = str.toCharArray();
		process1(chs, 0);
	}

	// process1打印字符串全排列，允许重复，会产生n!种情况
	// 比如acc的全排列，会产生3!=6个字符串
	public static void process1(char[] chs, int i) { // 可选元素的范围：下标从i到之后的所有元素中 放置位置：选择元素放到位置i上
		if (i == chs.length) {
			System.out.println(String.valueOf(chs));
		}
		for (int j = i; j < chs.length; j++) {
			swap(chs, i, j);
			process1(chs, i + 1); // 继续递归
			swap(chs, i, j);
		}
	}

	public static void printAllPermutations2(String str) {
		char[] chs = str.toCharArray();
		process2(chs, 0);
	}

	// process2打印字符串全排列，不允许重复
	// 比如求acc的全排列，会产生acc cac cca 三个字符串
	// hashset用于同层去重
	public static void process2(char[] chs, int i) { // 放置位置：i 可选元素范围：i以及之后的下标对应元素
		if (i == chs.length) {
			System.out.println(String.valueOf(chs));
		}
		// 每次调用process2都会产生一个新的hashset->hashset用于同一层上的去重
		HashSet<Character> set = new HashSet<>(); // hashset并不存储结果，而是去除掉已经重复的子串
		for (int j = i; j < chs.length; j++) {
			if (!set.contains(chs[j])) { // hashset中已经有的
				set.add(chs[j]);
				swap(chs, i, j); // 选择元素放到位置i上
				process2(chs, i + 1);
				swap(chs, i, j);
			}
		}
	}

	public static void swap(char[] chs, int i, int j) {
		char tmp = chs[i];
		chs[i] = chs[j];
		chs[j] = tmp;
	}

	public static void main(String[] args) {
		String test1 = "abc";
		printAllPermutations1(test1);
		System.out.println("======");
		printAllPermutations2(test1);
		System.out.println("======");

		String test2 = "acc";
		printAllPermutations1(test2);
		System.out.println("======");
		printAllPermutations2(test2);
		System.out.println("======");
	}

}
